File: python/141_hasCycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        first = slow = head
        
        while first and slow and first.next:
            first = first.next.next
            slow = slow.next

            if first is slow:
                return True
        
        return False

File: 141_hasCycle.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool hasCycle(struct ListNode *head) {
    struct ListNode *fast = head;
    struct ListNode *slow = head;
    
    while(fast && slow && fast->next) {
        fast = fast->next->next;
        slow = slow->next;
        
        if (fast == slow)
            return true;
    }
    
    return false;
}
